File: electron/providers/tools/read.ts
/**
 * Read Tool - Read file contents
 */

import * as fs from "fs";
import * as path from "path";

export interface ReadInput {
  file_path: string;
  offset?: number;
  limit?: number;
}

export interface ReadOutput {
  content: string;
  lineCount: number;
}

export function readTool(input: ReadInput, cwd: string): ReadOutput {
  const filePath = path.isAbsolute(input.file_path)
    ? input.file_path
    : path.resolve(cwd, input.file_path);

  if (!fs.existsSync(filePath)) {
    throw new Error(`File not found: ${filePath}`);
  }

  const stats = fs.statSync(filePath);
  if (stats.isDirectory()) {
    throw new Error(`Path is a directory, not a file: ${filePath}`);
  }

  const content = fs.readFileSync(filePath, "utf-8");
  const lines = content.split("\n");

  const offset = input.offset ?? 0;
  const limit = input.limit ?? lines.length;

  const selectedLines = lines.slice(offset, offset + limit);

  // Format with line numbers (1-indexed)
  const numberedLines = selectedLines.map((line, i) => {
    const lineNum = offset + i + 1;
    return `${String(lineNum).padStart(6, " ")}\t${line}`;
  });

  return {
    content: numberedLines.join("\n"),
    lineCount: lines.length,
  };
}

File: src-tauri/src/providers/tools/read.rs
use std::fs;
use std::path::Path;

use anyhow::{bail, Result};

/// Read file contents, returning numbered lines.
///
/// * `file_path` - Path to the file (absolute or relative to `cwd`).
/// * `cwd` - Current working directory for resolving relative paths.
/// * `offset` - Zero-indexed line to start reading from.
/// * `limit` - Maximum number of lines to return.
///
/// Returns a string with `cat -n` style numbered output.
pub fn read_tool(
    file_path: &str,
    cwd: &str,
    offset: Option<usize>,
    limit: Option<usize>,
) -> Result<String> {
    let path = if Path::new(file_path).is_absolute() {
        file_path.to_string()
    } else {
        Path::new(cwd).join(file_path).to_string_lossy().to_string()
    };

    let p = Path::new(&path);

    if !p.exists() {
        bail!("File not found: {}", path);
    }
    if p.is_dir() {
        bail!("Path is a directory, not a file: {}", path);
    }

    let content = fs::read_to_string(&path)?;
    let lines: Vec<&str> = content.split('\n').collect();

    let offset = offset.unwrap_or(0);
    let limit = limit.unwrap_or(lines.len());

    let end = std::cmp::min(offset + limit, lines.len());
    let selected = &lines[offset..end];

    // Format with line numbers (1-indexed)
    let numbered: Vec<String> = selected
        .iter()
        .enumerate()
        .map(|(i, line)| {
            let line_num = offset + i + 1;
            format!("{:>6}\t{}", line_num, line)
        })
        .collect();

    Ok(numbered.join("\n"))
}
